### PR TITLE
[RFC] Use JS heap instead of linear memory

### PR DESCRIPTION
**Pull request by wingo**
**This PR is an RFC; it is not actually a request to pull.**

This PR changes Schism to allocate its objects on an external heap, via call-outs to run-time imports.

Currently, references from wasm to external values is by "handle" -- an index into a table.  In the future this will be refactored to use the "anyref" facility.  In some ideal future, instead we use GC types, so that we can avoid callouts to the runtime.

The benchmark times are terrible; about 35 seconds on my machine.  By way of comparison, master is more like 4.  Bootstrap is not a great benchmark though; I need to work on a different benchmark that doesn't change as the compiler changes.

Based on #69

### CODE REVIEW

File: rt/rt.mjs [hunk 1]
```diff
@@ -14,114 +14,248 @@
...
-function extract_tag(value) {
-    return value & ((1 << TAG_SIZE) - 1);
+export class SmallInteger {
+    constructor(value) {
+        this.value = value | 0;
+    }
+    toJS(heap, map) {
+        return this.value;
+    }
+    copy(old_heap, new_heap, map) {
```
Comment (eholk) [inline rt/rt.mjs]: I like having each type implement its own copy function.

**Comment by wingo** [comment]
Hi again!  Thanks for the feedback :)

Regarding performance, it's terrible.  For a version of https://github.com/WebAssembly/benchmarks/pull/3 using this version of eval, an eval(fib(25)) allocates something like 2.5 GB of memory and then aborts:

```
$ time ~/src/node/out/Release/node --experimental-modules /tmp/schism-eval-fib/run.mjs
(node:11239) ExperimentalWarning: The ESM module loader is experimental.

<--- Last few GCs --->

[11239:0x557324692430]   194541 ms: Mark-sweep 2051.2 (2054.4) -> 2050.8 (2055.4) MB, 2887.0 / 0.0 ms  (average mu = 0.119, current mu = 0.047) allocation failure scavenge might not succeed
[11239:0x557324692430]   197483 ms: Mark-sweep 2052.2 (2055.4) -> 2051.8 (2056.4) MB, 2800.4 / 0.0 ms  (average mu = 0.086, current mu = 0.048) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x557322c1afb9]
Security context: 0x1a99c5ec0911 <JSObject>
    1: newPair(aka newPair) [0xc7cb569c009] [file:///tmp/schism-eval-fib/rt.mjs:~201] [pc=0x3633aedc4b3b](this=0x3675715804d1 <undefined>,72,2)
    2: WasmToJsFrame [pc: 0x301c20200d9d]
    3: WASM [wasm-000157d2], function #45 ('cons'), pc=0x301c202060ad (+0x2d), pos=2698 (+7)
    4: WASM [wasm-000157d2], function #155 ('%eval'), pc=0x301c20212309 (+0x229),...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

Writing Node.js report to file: report.20190723.131320.11239.0.001.json
Node.js report completed
 1: 0x5573222168c0 node::Abort() [/home/wingo/src/node/out/Release/node]
 2: 0x5573222193fd node::OnFatalError(char const*, char const*) [/home/wingo/src/node/out/Release/node]
 3: 0x55732239a242 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/home/wingo/src/node/out/Release/node]
 4: 0x55732239a5bf v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/home/wingo/src/node/out/Release/node]
 5: 0x5573225a13e5  [/home/wingo/src/node/out/Release/node]
 6: 0x5573225adc10 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/home/wingo/src/node/out/Release/node]
 7: 0x5573225aec36 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/home/wingo/src/node/out/Release/node]
 8: 0x5573225b184c v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [/home/wingo/src/node/out/Release/node]
 9: 0x557322578187 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType) [/home/wingo/src/node/out/Release/node]
10: 0x557322887478 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/home/wingo/src/node/out/Release/node]
11: 0x557322c1afb9  [/home/wingo/src/node/out/Release/node]
Aborted

real	3m20.681s
user	4m53.806s
sys	0m10.187s
```

The problem is that we don't know what's on the stack, so we can't GC in any sensible way.  We'd have to maintain a shadow stack in that case, which would be sad.  I think the anyref solution is way better; it at least completes in 4 or 5 seconds for this case, and stays under 40 MB of memory usage; still a lot of course, but there's lots of allocation happening absent string literals.

As far as where to go.  Considering that anyref is shipping in all browsers behind a flag, and that tail calls are close to happening, and that perf with anyref is good, and that with anyref we get JS interop and GC and no problems with cycles -- I think Schism should go ahead and target these experimental proposals (anyref + tail calls) in its mainline branch, via #72.  Browsers are still a little bit behind in the sense that the features are flagged but this will fix itself soon enough. WDYT?

**Comment by eholk** [comment]
This sounds good to me. It's hard to argue with getting basically the same performance, but much better memory usage and JS interop for free.

As far as the shadow stack goes, originally, I was planning to insert a CPS pass early in the compiler. Then garbage collection is just doing a deep copy of the continuation. I expect that has even more downsides than a shadow stack, but it also gives us `call/cc` pretty much for free.

I'll take a look at #72 and try to get that merged.

**Comment by wingo** [comment]
Thanks for all the feedback!  Closing this one as it was in practice merged via #72.

**Pull request closed without merging**